spec/models/ability_spec.rb

```diff
@@ -89,11 +89,103 @@
       let(:other_self_organized_track) { create(:track, :self_organized) }
 
       it{ should be_able_to(:manage, user) }
-
       it{ should be_able_to(:manage, registration_public) }
       it{ should be_able_to(:manage, registration_not_public) }
-      it{ should_not be_able_to(:new, Registration.new(conference_id: conference_with_closed_registration.id))}
-      it{ should_not be_able_to(:create, Registration.new(conference_id: conference_with_closed_registration.id))}
+
+      # Test for user can register or not
+      context 'when user is not a speaker with event confirmed' do
+        let(:conference) { create(:conference) }
+
+        context 'when the registration is closed' do
+          let(:conference_with_closed_registration) { create(:conference) }
+
+          it{ should_not be_able_to(:new, Registration.new(conference: conference_with_closed_registration)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference_with_closed_registration)) }
+        end
+
+        context 'when the registration period is not set' do
+          it{ should_not be_able_to(:new, Registration.new(conference: conference)) }
+          it{ should_not be_able_to(:create, Registration.new(conference: conference)) }
+        end
+
+        context 'when user has already registered' do
```

Comment:
Let's also test this for when the user is a speaker with confirmed event.

Reply:
This test is already there.